### ISSUE TITLE
VS Code opens file from its cache without warning that it is showing content that does not match the file opened from disk

### ISSUE THREAD

**Issue by margaretselzer** [microsoft/vscode#169905]
Type: <b>Bug</b>

I will describe what I just did prior to getting here, but I am not 100% sure if this is guaranteed to reproduce the issue as I did not try repeating the steps just to see. 

I finished editing my code in a branch (old_branch) which I sent for review in a pull request (PR). Then I branched out from this branch (new_branch) and continued development of a new feature. When I got some comments on my PR, I did some edits in the file that I got comments on and committed these changes in new_branch. Then I switched branches (using an external program, not in VS Code) and went back to old_branch to apply the same changes in the same file. 

Nevertheless I noticed that when I opened the file in old_branch, I got served a strange mixed version of the file, mixing in the latest changes from new_branch but keeping some stuff from the old_branch version too. Then I closed VS Code and went to the source file in Windows Explorer. Opened it in Notepad++ and I saw the correct version. Then opened it in VS Code and I saw the mixed-up version. I tried restarting everything, closing the file, opening the file, disabling/enabling git in VS Code, nothing happened.

Finally in my frustration I did some change to the file and tried to save it when I finally got a message saying: 
"Failed to save 'RuleEvaluator.h': The content of the file is newer. Please compare your version with the file contents or overwrite the content of the file with your changes."

So I realised that VS Code is not totally lost as it knows it handles a file not matching the contents of the file on disk. The only problem is that it serves up this version from its cache (or I do not know from where) without any warning whatsoever and there is no way to open up the actual file from disk (or I at least do not know how) in VS Code. So if I did not know what the contents of the file should be, I would never realise that this was wrong.

In this example I have no idea why a cached version would be kept by VS code since I saved the file and even committed the changes. In any case, these kind of saved cache versions of a file should be handled similarly to how MS Office handles a "recovery version" of a file that it automatically saved before a crash. I would not want to see though that same panel with "saved versions" opened up as in MS Office. It is totally enough if there exists only one "recovery version" of any given file. We use git after all for all the rest...

Update: As it might make a difference, I just want to clarify that I did not use the built-in source control/git in VS Code. I normally use GitExtensions for that.

VS Code version: Code 1.74.2 (e8a3071ea4344d9d48ef8a4df2c097372b0c5161, 2022-12-20T10:29:14.590Z)
OS version: Windows_NT x64 10.0.19044
Modes:
Sandboxed: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i5-1145G7 @ 2.60GHz (8 x 2611)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|undefined|
|Memory (System)|31.69GB (13.60GB free)|
|Process Argv|--crash-reporter-id edb703fb-e2d6-40c7-8c52-fa023c25e7c3|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (10)</summary>

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|13.3.1
doxdocgen|csc|1.4.0
better-cpp-syntax|jef|1.17.2
cpptools|ms-|1.13.8
platformio-ide|pla|2.5.5
annotator|ryu|1.0.0
sonarlint-vscode|Son|3.13.0
code-spell-checker|str|2.12.0
gitblame|wad|10.1.0
vscode-todo-highlight|way|1.0.5

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vscaat:30438848
vsclangdc:30486549
c4g48928:30535728
dsvsc012cf:30540253
azure-dev_surveyonecf:30548226
pyindex848cf:30577861
nodejswelcome1cf:30587006
2e4cg342:30602488
gswce1:30612156
iaj6b796:30613358
f6dab269:30613381
fim-prod:30623723

```

</details>

**Comment by bpasero** [comment]
We always fetch the contents from disk unless the file is dirty, so I suspect in this case here the file was dirty when it showed the weird mixed contents?

**Comment by margaretselzer** [comment]
As I wrote, I saved and committed the file before changing branches, so it was not - should not have been - dirty.

**Comment by bpasero** [comment]
Without some steps to reproduce this reliably I believe I cannot drill into what happened. Is the file system any special?

**Comment by margaretselzer** [comment]
No, nothing special about the file system. The only thing that maybe causing an issue is that the repository is in a synced OneDrive folder - though the way I use it, it is really a one-way backup to the cloud only, or at least it should be so. (There was a period of time when I was missing some access rights to our Azure repo and tried to have some safety with using OneDrive instead. And then it stayed like that.)

Well, if the steps I wrote down in the issue do not reproduce the error on your side and you have no idea what could have caused this, then I understand if you cannot proceed. 

I am experiencing some other weird behavior also on my desktop which might be also at play: branches and files/folders that I deleted reappear suddenly out of nowhere. The day before yesterday I was preparing to commit some changes and suddenly git told me I was in the middle of a rebase with files that I deleted in another branch reappearing. When I aborted the rebase all my latest changes were gone with it. Luckily VS Code undo saved me there as all the impacted files were open at the time. But as I cannot unambiguously link this to any particular software I did not report it anywhere. I just wrote it down in case you have this issue coming in from somewhere else as well in which case it might be useful.
